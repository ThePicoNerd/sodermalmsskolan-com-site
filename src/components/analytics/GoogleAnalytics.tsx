import Head from "next/head";
import React, { FunctionComponent, useEffect } from "react";
import gtag from "../../lib/analytics/gtag";

export interface GoogleAnalyticsProps {
  trackingId: string;
  appName: string;
  appVersion: string;
}

/**
 * Google Analytics tracking snippet, as a React component. Handy, right?
 *
 * @param {React.PropsWithChildren<GoogleAnalyticsProps>} props Props.
 *
 * @returns {React.ReactElement} The rendered code.
 */
const GoogleAnalytics: FunctionComponent<GoogleAnalyticsProps> = ({
  trackingId,
  appName,
  appVersion,
}) => {
  useEffect(() => {
    // Required for GA to function, as per the snippet.
    gtag("js", new Date());
    gtag("config", trackingId);

    // Not required, but only specified during initialization.
    gtag("set", "appName", appName);
    gtag("set", "appVersion", appVersion);
  }, [appName, appVersion, trackingId]);

  return (
    <Head>
      {/* Global site tag (gtag.js) - Google Analytics */}
      <script
        async
        src={`https://www.googletagmanager.com/gtag/js?id=${trackingId}`}
      />
      <script
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{
          __html: `window.dataLayer = window.dataLayer || [];
              function gtag(){dataLayer.push(arguments);}`,
        }}
      />
    </Head>
  );
};

export default GoogleAnalytics;
