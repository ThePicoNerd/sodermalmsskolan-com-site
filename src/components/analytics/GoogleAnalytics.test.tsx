import { render } from "@testing-library/react";
import React from "react";
import ANALYTICS_ID from "../../lib/analytics/constants";
import GoogleAnalytics from "./GoogleAnalytics";

describe("<GoogleAnalytics /> test", () => {
  it("should return the correct tags", () => {
    const result = render(<GoogleAnalytics trackingId={ANALYTICS_ID} />);

    expect(result.asFragment()).toMatchSnapshot();
  });
});
