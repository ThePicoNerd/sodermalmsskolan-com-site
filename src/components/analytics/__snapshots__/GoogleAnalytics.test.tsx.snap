// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoogleAnalytics /> test should return the correct tags 1`] = `
<DocumentFragment>
  <script
    async=""
    src="https://www.googletagmanager.com/gtag/js?id=G-P3DFFXTCY0"
  />
  <script>
    window.dataLayer = window.dataLayer || [];
              function gtag(){dataLayer.push(arguments);}
  </script>
</DocumentFragment>
`;
